### PR TITLE
[Feature] `Files()` support parquet nested type

### LINKED ISSUE #30613: `Files()` support nested types
## Feature request
The table function `files()` still dose not support nested types, like `MAP` and `ARRAY`. It's highly needed.

**Is your feature request related to a problem? Please describe.**

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by rickif**
Fixes #30613 

## What type of PR is this:
This PR adds the support of parquet nested types `LIST`/`MAP` to table function `files()`

Here's an example.
```
MySQL [db0]> create table tbl_nested as select * from  files("path"="hdfs://127.0.0.1:9000/nested-type-test/nested.parquet","format"="parquet");
Query OK, 3 rows affected (1.604 sec)
{'label':'insert_2317ffcb-5603-11ee-acb3-66109f72cdee', 'status':'VISIBLE', 'txnId':'2'}

MySQL [db0]> desc tbl_nested;
+-------------------------+-------------------------------------------------+------+-------+---------+-------+
| Field                   | Type                                            | Null | Key   | Default | Extra |
+-------------------------+-------------------------------------------------+------+-------+---------+-------+
| col_int                 | bigint                                          | YES  | true  | NULL    |       |
| col_list_int            | array<int>                                      | YES  | false | NULL    |       |
| col_list_list_int       | array<array<int>>                               | YES  | false | NULL    |       |
| col_map_string_int      | map<varchar(1048576),int>                       | YES  | false | NULL    |       |
| col_map_map_string_int  | map<varchar(1048576),map<varchar(1048576),int>> | YES  | false | NULL    |       |
| col_list_map_string_int | array<map<varchar(1048576),int>>                | YES  | false | NULL    |       |
| col_map_string_list_int | map<varchar(1048576),array<int>>                | YES  | false | NULL    |       |
+-------------------------+-------------------------------------------------+------+-------+---------+-------+
7 rows in set (0.005 sec)

```
The python script to generate `nested.parquet` is added in this PR.

- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:

- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:disappointed: **fail** : 2 / 8 (25.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/runtime/types.h|0|6|00.00%|[120, 121, 122, 123, 124, 127]|
|:large_blue_circle:|src/exec/parquet_reader.cpp|2|2|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:disappointed: **fail** : 0 / 14 (00.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/TableFunctionTable.java|0|1|00.00%|[248]|
|:large_blue_circle:|com/starrocks/catalog/Type.java|0|13|00.00%|[1313, 1314, 1315, 1317, 1318, 1321, 1322, 1323, 1326, 1327, 1328, 1329, 1332]|

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:disappointed: **fail** : 0 / 14 (00.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/TableFunctionTable.java|0|1|00.00%|[248]|
|:large_blue_circle:|com/starrocks/catalog/Type.java|0|13|00.00%|[1313, 1314, 1315, 1317, 1318, 1321, 1322, 1323, 1326, 1327, 1328, 1329, 1332]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:heart_eyes: **pass** : 101 / 114 (88.60%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/runtime/types.h|0|6|00.00%|[120, 121, 122, 123, 124, 125]|
|:large_blue_circle:|src/exec/parquet_schema_builder.cpp|100|107|93.46%|[55, 57, 70, 92, 101, 105, 107]|
|:large_blue_circle:|src/exec/parquet_reader.cpp|1|1|100.00%|[]|

**Comment by rickif** [comment]
@mergifyio rebase

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:disappointed: **fail** : 0 / 14 (00.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/TableFunctionTable.java|0|1|00.00%|[248]|
|:large_blue_circle:|com/starrocks/catalog/Type.java|0|13|00.00%|[1313, 1314, 1315, 1317, 1318, 1321, 1322, 1323, 1326, 1327, 1328, 1329, 1332]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:heart_eyes: **pass** : 101 / 114 (88.60%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/runtime/types.h|0|6|00.00%|[120, 121, 122, 123, 124, 125]|
|:large_blue_circle:|src/exec/parquet_schema_builder.cpp|100|107|93.46%|[55, 57, 70, 92, 101, 105, 107]|
|:large_blue_circle:|src/exec/parquet_reader.cpp|1|1|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:disappointed: **fail** : 12 / 17 (70.59%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/TableFunctionTable.java|0|4|00.00%|[248, 249, 250, 252]|
|:large_blue_circle:|com/starrocks/catalog/Type.java|12|13|92.31%|[1332]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:heart_eyes: **pass** : 101 / 114 (88.60%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/runtime/types.h|0|6|00.00%|[120, 121, 122, 123, 124, 125]|
|:large_blue_circle:|src/exec/parquet_schema_builder.cpp|100|107|93.46%|[55, 57, 70, 92, 101, 105, 107]|
|:large_blue_circle:|src/exec/parquet_reader.cpp|1|1|100.00%|[]|

**Comment by rickif** [comment]
@mergifyio rebase

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:disappointed: **fail** : 12 / 17 (70.59%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/TableFunctionTable.java|0|4|00.00%|[284, 285, 286, 288]|
|:large_blue_circle:|com/starrocks/catalog/Type.java|12|13|92.31%|[1332]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:heart_eyes: **pass** : 101 / 114 (88.60%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/runtime/types.h|0|6|00.00%|[120, 121, 122, 123, 124, 125]|
|:large_blue_circle:|src/exec/parquet_schema_builder.cpp|100|107|93.46%|[55, 57, 70, 92, 101, 105, 107]|
|:large_blue_circle:|src/exec/parquet_reader.cpp|1|1|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:disappointed: **fail** : 12 / 17 (70.59%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/TableFunctionTable.java|0|4|00.00%|[284, 285, 286, 288]|
|:large_blue_circle:|com/starrocks/catalog/Type.java|12|13|92.31%|[1332]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:heart_eyes: **pass** : 101 / 114 (88.60%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/runtime/types.h|0|6|00.00%|[120, 121, 122, 123, 124, 125]|
|:large_blue_circle:|src/exec/parquet_schema_builder.cpp|100|107|93.46%|[55, 57, 70, 92, 101, 105, 107]|
|:large_blue_circle:|src/exec/parquet_reader.cpp|1|1|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:disappointed: **fail** : 12 / 17 (70.59%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/TableFunctionTable.java|0|4|00.00%|[284, 285, 286, 288]|
|:large_blue_circle:|com/starrocks/catalog/Type.java|12|13|92.31%|[1332]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:heart_eyes: **pass** : 101 / 114 (88.60%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/runtime/types.h|0|6|00.00%|[120, 121, 122, 123, 124, 125]|
|:large_blue_circle:|src/exec/parquet_schema_builder.cpp|100|107|93.46%|[55, 57, 70, 92, 101, 105, 107]|
|:large_blue_circle:|src/exec/parquet_reader.cpp|1|1|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:disappointed: **fail** : 12 / 17 (70.59%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/TableFunctionTable.java|0|4|00.00%|[284, 285, 286, 288]|
|:large_blue_circle:|com/starrocks/catalog/Type.java|12|13|92.31%|[1332]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:heart_eyes: **pass** : 101 / 114 (88.60%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/runtime/types.h|0|6|00.00%|[120, 121, 122, 123, 124, 125]|
|:large_blue_circle:|src/exec/parquet_schema_builder.cpp|100|107|93.46%|[55, 57, 70, 92, 101, 105, 107]|
|:large_blue_circle:|src/exec/parquet_reader.cpp|1|1|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:disappointed: **fail** : 12 / 17 (70.59%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/TableFunctionTable.java|0|4|00.00%|[284, 285, 286, 288]|
|:large_blue_circle:|com/starrocks/catalog/Type.java|12|13|92.31%|[1332]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:heart_eyes: **pass** : 101 / 115 (87.83%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/runtime/types.h|0|7|00.00%|[113, 120, 121, 122, 123, 124, 125]|
|:large_blue_circle:|src/exec/parquet_schema_builder.cpp|100|107|93.46%|[55, 57, 70, 92, 101, 105, 107]|
|:large_blue_circle:|src/exec/parquet_reader.cpp|1|1|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:disappointed: **fail** : 12 / 17 (70.59%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/TableFunctionTable.java|0|4|00.00%|[284, 285, 286, 288]|
|:large_blue_circle:|com/starrocks/catalog/Type.java|12|13|92.31%|[1332]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:heart_eyes: **pass** : 101 / 115 (87.83%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/runtime/types.h|0|7|00.00%|[113, 120, 121, 122, 123, 124, 125]|
|:large_blue_circle:|src/exec/parquet_schema_builder.cpp|100|107|93.46%|[55, 57, 70, 92, 101, 105, 107]|
|:large_blue_circle:|src/exec/parquet_reader.cpp|1|1|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:disappointed: **fail** : 20 / 28 (71.43%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/TableFunctionTable.java|0|7|00.00%|[283, 284, 285, 290, 291, 292, 294]|
|:large_blue_circle:|com/starrocks/catalog/Type.java|20|21|95.24%|[1344]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:heart_eyes: **pass** : 101 / 115 (87.83%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/runtime/types.h|0|7|00.00%|[113, 120, 121, 122, 123, 124, 125]|
|:large_blue_circle:|src/exec/parquet_schema_builder.cpp|100|107|93.46%|[55, 57, 70, 92, 101, 105, 107]|
|:large_blue_circle:|src/exec/parquet_reader.cpp|1|1|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:disappointed: **fail** : 20 / 28 (71.43%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/TableFunctionTable.java|0|7|00.00%|[283, 284, 285, 290, 291, 292, 294]|
|:large_blue_circle:|com/starrocks/catalog/Type.java|20|21|95.24%|[1344]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:heart_eyes: **pass** : 101 / 115 (87.83%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/runtime/types.h|0|7|00.00%|[113, 120, 121, 122, 123, 124, 125]|
|:large_blue_circle:|src/exec/parquet_schema_builder.cpp|100|107|93.46%|[55, 57, 70, 92, 101, 105, 107]|
|:large_blue_circle:|src/exec/parquet_reader.cpp|1|1|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:disappointed: **fail** : 20 / 28 (71.43%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/TableFunctionTable.java|0|7|00.00%|[283, 284, 285, 290, 291, 292, 294]|
|:large_blue_circle:|com/starrocks/catalog/Type.java|20|21|95.24%|[1344]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:heart_eyes: **pass** : 101 / 115 (87.83%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/runtime/types.h|0|7|00.00%|[113, 120, 121, 122, 123, 124, 125]|
|:large_blue_circle:|src/exec/parquet_schema_builder.cpp|100|107|93.46%|[55, 57, 70, 92, 101, 105, 107]|
|:large_blue_circle:|src/exec/parquet_reader.cpp|1|1|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:heart_eyes: **pass** : 25 / 28 (89.29%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/TableFunctionTable.java|5|7|71.43%|[285, 292]|
|:large_blue_circle:|com/starrocks/catalog/Type.java|20|21|95.24%|[1344]|

**Comment by rickif** [comment]
@mergifyio rebase

File: fe/fe-core/src/main/java/com/starrocks/catalog/TableFunctionTable.java [hunk 1]
```diff
@@ -281,15 +280,18 @@ private List<Column> getFileSchema() throws DdlException {
             throw new DdlException("failed to get file schema", e);
         }
 
+        TStatusCode code = TStatusCode.findByValue(result.status.statusCode);
+        if (code != TStatusCode.OK) {
+            throw new DdlException("failed to get file schema: " + result.status.errorMsgs);
```
Comment (chaoyli) [inline fe/fe-core/src/main/java/com/starrocks/catalog/TableFunctionTable.java:285]: output the file name to better trace.

File: fe/fe-core/src/main/java/com/starrocks/catalog/TableFunctionTable.java [hunk 2]
```diff
@@ -281,15 +280,18 @@ private List<Column> getFileSchema() throws DdlException {
...
+
         List<Column> columns = new ArrayList<>();
         for (PSlotDescriptor slot : result.schema) {
-
-            List<PTypeNode> types = slot.slotType.types;
-            if (types.size() != 1) {
-                throw new DdlException("non-scalar type is not supported: " + slot.colName);
+            Type tp = Type.fromProtobuf(slot.slotType, 0);
+            if (tp == null) {
+                throw new DdlException("unsupported type:" + slot.slotType.toString() + " ,column name: " + slot.colName);
```
Comment (chaoyli) [inline fe/fe-core/src/main/java/com/starrocks/catalog/TableFunctionTable.java:292]: output the file name to better trace.

File: be/test/exec/test_data/parquet_data/generate_nested.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
+# Copyright 2021-present StarRocks, Inc. All rights reserved.
```
[the remark this answers is not in the document]
Comment (rickif) [reply be/test/exec/test_data/parquet_data/generate_nested.py:1]: We already have many parquet files. However, when I need to add a new test case with the existing parquet file, I have no idea how to edit these files. With the python script, we could update the data file on demand in the future.

File: fe/fe-core/src/main/java/com/starrocks/catalog/TableFunctionTable.java (hunk 2 above)
Comment (rickif) [reply fe/fe-core/src/main/java/com/starrocks/catalog/TableFunctionTable.java]: Already removed this log.

**Comment by rickif** [comment]
@mergifyio rebase

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:disappointed: **fail** : 20 / 27 (74.07%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/TableFunctionTable.java|0|5|00.00%|[283, 284, 285, 290, 291]|
|:large_blue_circle:|com/starrocks/catalog/Type.java|20|22|90.91%|[1345, 1346]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:heart_eyes: **pass** : 101 / 115 (87.83%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/runtime/types.h|0|7|00.00%|[113, 120, 121, 122, 123, 124, 125]|
|:large_blue_circle:|src/exec/parquet_schema_builder.cpp|100|107|93.46%|[55, 57, 70, 92, 101, 105, 107]|
|:large_blue_circle:|src/exec/parquet_reader.cpp|1|1|100.00%|[]|

**Review by chaoyli — DISMISSED** [review]

**Comment by rickif** [comment]
@mergifyio rebase

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:heart_eyes: **pass** : 20 / 25 (80.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/TableFunctionTable.java|0|3|00.00%|[311, 312, 317]|
|:large_blue_circle:|com/starrocks/catalog/Type.java|20|22|90.91%|[1345, 1346]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:heart_eyes: **pass** : 101 / 115 (87.83%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/runtime/types.h|0|7|00.00%|[113, 120, 121, 122, 123, 124, 125]|
|:large_blue_circle:|src/exec/parquet_schema_builder.cpp|100|107|93.46%|[55, 57, 70, 92, 101, 105, 107]|
|:large_blue_circle:|src/exec/parquet_reader.cpp|1|1|100.00%|[]|

**Review by meegoo — APPROVED** [review]

**Review by chaoyli — APPROVED** [review]

**Comment by rickif** [comment]
@mergifyio backport branch-3.2

**Pull request merged**